[libpas] Clean up enumerator memory crawling failure handling code in pas_report_crash.c

zakariaridouh opened the pull request:
#### 6f816fb9ce69512c678296cdf3b7839f3ae42d3c
<pre>
[libpas] Clean up enumerator memory crawling failure handling code in pas_report_crash.c
<a href="https://bugs.webkit.org/show_bug.cgi?id=300746">https://bugs.webkit.org/show_bug.cgi?id=300746</a>
<a href="https://rdar.apple.com/162639347">rdar://162639347</a>

Reviewed by Dan Hecht and Brandon Stewart.

After 301166@main, we have a lot of duplicate code essentially destroying the enumerator and returning failure state.

We can clean this up by introducing a fail function.

* Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
(pas_enumerator_destroy_and_fail):
(pas_report_crash_extract_pgm_failure):

Canonical link: <a href="https://commits.webkit.org/301561@main">https://commits.webkit.org/301561@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/34c0b1d1e35c76c5dad9fe526c6154afdb7337ab

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126222 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45878 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36695 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133000 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77996 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f0c3ad15-634f-4bee-bf37-2359d1e6445d) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46542 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54424 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96099 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64211 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/25126dab-c774-4d9d-834a-d1a6a3344c3a) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129170 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37232 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112899 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76581 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/abc21378-d309-43b5-87a7-86282273d6a4) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36132 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31078 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76472 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/118302 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107014 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31307 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135701 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/124690 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52971 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40689 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104600 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53433 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109123 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104303 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26625 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49737 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28073 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50354 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52871 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/157737 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52187 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39473 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55533 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53904 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

stwrt commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
```diff
@@ -72,6 +72,12 @@ static PAS_ALWAYS_INLINE kern_return_t PAS_WARN_UNUSED_RETURN pas_update_report_
     return KERN_SUCCESS;
 }
 
+static PAS_ALWAYS_INLINE kern_return_t PAS_WARN_UNUSED_RETURN pas_enumerator_destroy_and_fail(pas_enumerator* enumerator)
```
Why do we need `PAS_WARN_UNUSED_RETURN`?

All the returns seem to be used unless I am misunderstanding this macro.

stwrt replied:
Nevermind, I was thinking about this macro incorrectly.

zakariaridouh committed:
[libpas] Clean up enumerator memory crawling failure handling code in pas_report_crash.c
https://bugs.webkit.org/show_bug.cgi?id=300746
rdar://162639347

Reviewed by Dan Hecht and Brandon Stewart.

After 301166@main, we have a lot of duplicate code essentially destroying the enumerator and returning failure state.

We can clean this up by introducing a fail function.

* Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
(pas_enumerator_destroy_and_fail):
(pas_report_crash_extract_pgm_failure):

Canonical link: https://commits.webkit.org/301561@main
